In [1]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd


In [2]:
# Add the encoding='latin1' parameter
df = pd.read_csv(r"C:\Users\DeLL\Desktop\Car_Price_Prediction\car_purchasing.csv", encoding='latin1')

In [3]:
df.head(2)

customer name                                    customer e-mail   country  \
0  Martina Avila  cubilia.Curae.Phasellus@quisaccumsanconvallis.edu  Bulgaria   
1  Harlan Barnes                                eu.dolor@diam.co.uk    Belize   

   gender        age  annual Salary  credit card debt    net worth  \
0       0  41.851720    62812.09301      11609.380910  238961.2505   
1       0  40.870623    66646.89292       9572.957136  530973.9078   

   car purchase amount  
0          35321.45877  
1          45115.52566

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   customer name        500 non-null    object 
 1   customer e-mail      500 non-null    object 
 2   country              500 non-null    object 
 3   gender               500 non-null    int64  
 4   age                  500 non-null    float64
 5   annual Salary        500 non-null    float64
 6   credit card debt     500 non-null    float64
 7   net worth            500 non-null    float64
 8   car purchase amount  500 non-null    float64
dtypes: float64(5), int64(1), object(3)
memory usage: 35.3+ KB


In [5]:
df.isnull().sum()

customer name          0
customer e-mail        0
country                0
gender                 0
age                    0
annual Salary          0
credit card debt       0
net worth              0
car purchase amount    0
dtype: int64

In [8]:
df.duplicated().sum()

np.int64(0)

In [9]:
print(df['country'].unique())

['Bulgaria' 'Belize' 'Algeria' 'Cook Islands' 'Brazil' 'Liberia' 'Syria'
 'Czech Republic' 'Armenia' 'Somalia' 'Sint Maarten' 'Greenland'
 'Nicaragua' 'Palestine, State of' 'United Arab Emirates' 'Gabon'
 'Tokelau' 'Portugal' 'Chad' 'Iraq' 'Sudan' 'Angola' 'Nigeria'
 'Madagascar' 'Macedonia' 'Oman' 'Colombia' 'Namibia' 'Denmark'
 'Dominican Republic' 'Costa Rica' 'Botswana' 'Saint Pierre and Miquelon'
 'France' 'Senegal' 'Cambodia' 'Suriname' 'Turkey' 'Ethiopia'
 'French Guiana' 'Bhutan' 'Georgia' 'Mozambique' 'Nauru' 'Timor-Leste'
 'Djibouti' 'Argentina' 'Taiwan' 'Åland Islands'
 'Turks and Caicos Islands' 'Samoa' 'Vanuatu' 'Mongolia' 'Sri Lanka'
 'Guinea-Bissau' 'Egypt' 'Slovenia' 'Benin' 'Cape Verde' 'Paraguay' 'Laos'
 'Iceland' 'Viet Nam' 'Marshall Islands' 'Germany' 'Kyrgyzstan'
 'Mauritius' 'Gambia' 'Reunion' 'Jordan' 'Slovakia' 'Kiribati'
 'United States Minor Outlying Islands' 'Puerto Rico' 'Guinea' 'Guam'
 'China' 'Sierra Leone' 'Ecuador' 'Anguilla' 'Kazakhstan' 'Jersey' 'Maca

In [10]:
len(df['country'].unique())

211

In [11]:
df = df.drop(columns=['customer name', 'customer e-mail', 'country'])

In [12]:
df.head()

gender        age  annual Salary  credit card debt    net worth  \
0       0  41.851720    62812.09301      11609.380910  238961.2505   
1       0  40.870623    66646.89292       9572.957136  530973.9078   
2       1  43.152897    53798.55112      11160.355060  638467.1773   
3       1  58.271369    79370.03798      14426.164850  548599.0524   
4       1  57.313749    59729.15130       5358.712177  560304.0671   

   car purchase amount  
0          35321.45877  
1          45115.52566  
2          42925.70921  
3          67422.36313  
4          55915.46248

In [13]:
df.age = df.age.astype(int)

In [14]:
df.head()

gender  age  annual Salary  credit card debt    net worth  \
0       0   41    62812.09301      11609.380910  238961.2505   
1       0   40    66646.89292       9572.957136  530973.9078   
2       1   43    53798.55112      11160.355060  638467.1773   
3       1   58    79370.03798      14426.164850  548599.0524   
4       1   57    59729.15130       5358.712177  560304.0671   

   car purchase amount  
0          35321.45877  
1          45115.52566  
2          42925.70921  
3          67422.36313  
4          55915.46248

In [24]:
df['gender'].unique()

array([0, 1])

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   gender               500 non-null    int64  
 1   age                  500 non-null    int64  
 2   annual Salary        500 non-null    float64
 3   credit card debt     500 non-null    float64
 4   net worth            500 non-null    float64
 5   car purchase amount  500 non-null    float64
dtypes: float64(4), int64(2)
memory usage: 23.6 KB


In [16]:
# EDA Part is complete

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression


In [18]:
X =df.iloc[:, :-1]
y = df.iloc[:, -1]

In [26]:
print(X.dtypes)

gender                int64
age                   int64
annual Salary       float64
credit card debt    float64
net worth           float64
dtype: object


In [19]:
y.head()

0    35321.45877
1    45115.52566
2    42925.70921
3    67422.36313
4    55915.46248
Name: car purchase amount, dtype: float64

In [20]:
X.head()

gender  age  annual Salary  credit card debt    net worth
0       0   41    62812.09301      11609.380910  238961.2505
1       0   40    66646.89292       9572.957136  530973.9078
2       1   43    53798.55112      11160.355060  638467.1773
3       1   58    79370.03798      14426.164850  548599.0524
4       1   57    59729.15130       5358.712177  560304.0671

In [21]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [34]:
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPRegressor  # <-- IMPORT THE REGRESSOR
from sklearn.metrics import r2_score           # <-- IMPORT NEW METRIC

In [35]:
# Create the scaler
scaler = StandardScaler()

# Fit ONLY on the training data
X_train_scaled = scaler.fit_transform(X_train)

# Transform the test data
X_test_scaled = scaler.transform(X_test)


In [36]:
# 1. Create the ANN Regressor model
# We set max_iter=1000 to give it more time to train
ann_model = MLPRegressor(random_state=0, max_iter=1000)

# 2. Fit the model on the SCALED training data (This will work now)
ann_model.fit(X_train_scaled, y_train)

# 3. Predict on the SCALED test data
y_pred_ann = ann_model.predict(X_test_scaled)

# 4. Check the R² score (not accuracy)
# R² score of 1.0 is a perfect fit. 0 is a bad fit.
score_ann = r2_score(y_test, y_pred_ann)

print(f"ANN (MLPRegressor) R² Score: {score_ann}")

ANN (MLPRegressor) R² Score: -15.830414735287139


c:\Users\anmol\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


In [37]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

# 1. Create the model
lr_model = LinearRegression()

# 2. Fit on the original (unscaled) X_train
lr_model.fit(X_train, y_train)

# 3. Predict on the original (unscaled) X_test
y_pred_lr = lr_model.predict(X_test)

# 4. Get the R² score
score_lr = r2_score(y_test, y_pred_lr)
print(f"Linear Regression R² Score: {score_lr}")

Linear Regression R² Score: 0.9995462331275716


In [38]:
from sklearn.ensemble import RandomForestRegressor

# 1. Create the model
rf_model = RandomForestRegressor(random_state=0)

# 2. Fit on the original (unscaled) X_train
rf_model.fit(X_train, y_train)

# 3. Predict on the original (unscaled) X_test
y_pred_rf = rf_model.predict(X_test)

# 4. Get the R² score
score_rf = r2_score(y_test, y_pred_rf)
print(f"Random Forest R² Score: {score_rf}")

Random Forest R² Score: 0.9433798254097135


In [50]:
from sklearn.svm import SVR
from sklearn.metrics import r2_score

# 1. Create the SVR model
svr_model = SVR(kernel='rbf') # 'rbf' is a common and powerful kernel to start with

# 2. Fit on the SCALED training data
svr_model.fit(X_train_scaled, y_train)

# 3. Predict on the SCALED test data
y_pred_svr = svr_model.predict(X_test_scaled)

# 4. Get the R² score
score_svr = r2_score(y_test, y_pred_svr)
print(f"SVR R² Score: {score_svr}")

SVR R² Score: -7.511343922450564e-05


In [51]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import r2_score

# 1. Create the Gradient Boosting model
gbr_model = GradientBoostingRegressor(random_state=0)

# 2. Fit on the ORIGINAL training data
gbr_model.fit(X_train, y_train)

# 3. Predict on the ORIGINAL test data
y_pred_gbr = gbr_model.predict(X_test)

# 4. Get the R² score
score_gbr = r2_score(y_test, y_pred_gbr)
print(f"Gradient Boosting R² Score: {score_gbr}")

Gradient Boosting R² Score: 0.9691203405565131


In [52]:
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import r2_score

# 1. Create the KNN model (we'll start with k=5 neighbors)
knn_model = KNeighborsRegressor(n_neighbors=5)

# 2. Fit on the SCALED training data
knn_model.fit(X_train_scaled, y_train)

# 3. Predict on the SCALED test data
y_pred_knn = knn_model.predict(X_test_scaled)

# 4. Get the R² score
score_knn = r2_score(y_test, y_pred_knn)
print(f"K-Neighbors (KNN) R² Score: {score_knn}")

K-Neighbors (KNN) R² Score: 0.9028454313266264


In [39]:
from sklearn.metrics import mean_squared_error
import numpy as np

# You already have y_test and y_pred_lr from the Linear Regression model

# 1. Calculate Mean Squared Error (MSE)
mse = mean_squared_error(y_test, y_pred_lr)
print(f"Mean Squared Error (MSE): {mse}")

# 2. Calculate Root Mean Squared Error (RMSE)
rmse = np.sqrt(mse)
print(f"Root Mean Squared Error (RMSE): {rmse}")

Mean Squared Error (MSE): 54653.23162593494
Root Mean Squared Error (RMSE): 233.78030632612092


In [53]:
import numpy as np

# 1. Create your new example data
# These must be in the same order as your X columns
new_data = np.array([[0, 45, 60000, 3000, 250000]])

# 2. Use your 'lr_model' to predict
# (Your model is already trained)
prediction = lr_model.predict(new_data)

print(f"The predicted value is: {prediction[0]}")

The predicted value is: 37149.510102197695


c:\Users\anmol\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [42]:
X.head()

gender  age  annual Salary  credit card debt    net worth
0       0   41    62812.09301      11609.380910  238961.2505
1       0   40    66646.89292       9572.957136  530973.9078
2       1   43    53798.55112      11160.355060  638467.1773
3       1   58    79370.03798      14426.164850  548599.0524
4       1   57    59729.15130       5358.712177  560304.0671

In [43]:
import numpy as np

# 1. Create your new example data WITH 5 FEATURES
# [Gender, Age, Annual Salary, Credit Card Debt, Net Worth]
new_data = np.array([[0, 45, 60000, 3000, 250000]])  # <-- Added the '0' for Gender

# 2. Use your 'lr_model' to predict
# (Your model is already trained)
prediction = lr_model.predict(new_data)

print(f"The predicted value is: {prediction[0]}")

The predicted value is: 37149.510102197695


c:\Users\anmol\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [44]:
# 1. Create your new example data WITH 5 FEATURES
# [Gender, Age, Annual Salary, Credit Card Debt, Net Worth]
new_data = np.array([[0, 41, 62812.09301, 11609.380910,	238961.2505]])  # <-- Added the '0' for Gender

# 2. Use your 'lr_model' to predict
# (Your model is already trained)
prediction = lr_model.predict(new_data)

print(f"The predicted value is: {prediction[0]}")

The predicted value is: 35008.22575077996


c:\Users\anmol\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [46]:
y.head()

0    35321.45877
1    45115.52566
2    42925.70921
3    67422.36313
4    55915.46248
Name: car purchase amount, dtype: float64

In [47]:
# 1. Create your new example data WITH 5 FEATURES
# [Gender, Age, Annual Salary, Credit Card Debt, Net Worth]
new_data = np.array([[0, 40, 66646.89292, 9572.957136, 530973.9078]])  # <-- Added the '0' for Gender

# 2. Use your 'lr_model' to predict
# (Your model is already trained)
prediction = lr_model.predict(new_data)

print(f"The predicted value is: {prediction[0]}")

The predicted value is: 44807.438761289224


c:\Users\anmol\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [48]:
# 1. Create your new example data WITH 5 FEATURES
# [Gender, Age, Annual Salary, Credit Card Debt, Net Worth]
new_data = np.array([[1, 57, 59729.15130, 5358.712177, 560304.0671]])  # <-- Added the '0' for Gender

# 2. Use your 'lr_model' to predict
# (Your model is already trained)
prediction = lr_model.predict(new_data)

print(f"The predicted value is: {prediction[0]}")


The predicted value is: 56048.46974308331


c:\Users\anmol\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [49]:
# 1. Create your new example data WITH 5 FEATURES
# [Gender, Age, Annual Salary, Credit Card Debt, Net Worth]
new_data = np.array([[1, 58, 79370.03798, 14426.164850, 548599.0524]])  # <-- Added the '0' for Gender

# 2. Use your 'lr_model' to predict
# (Your model is already trained)
prediction = lr_model.predict(new_data)

print(f"The predicted value is: {prediction[0]}")

The predicted value is: 67576.61953071418


c:\Users\anmol\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


### So we can see that after applying multiple regression model the best R² score achieved by a model is 0.9995 by Linear Regression Model 

# so we can clearly say that Linear Relation model is the best fit model for this Car price pridection